### PR TITLE
Add eval like functionality to koto module

### PR DESCRIPTION
**Pull request by alisomay**
Tonight I had a bit of time and wanted to tackle 

- https://github.com/koto-lang/koto/issues/162

I'm making this a draft PR since I think a pre-review might be useful.

Do you think the implementation is in the right direction? Especially about error handling. 

As I've understood we don't want to panic the program due to a either compilation or runtime error coming from the evaluated script. On the other hand what should we do if the evaluated script panics the VM? I've included one case for that in the tests (commented out).

Thanks for the time and review 🙏  

After your initial comments, I'd be happy to make the necessary changes and conclude this as a real PR.

### CODE REVIEW

**Comment by irh** [comment]
Hi @alisomay, thanks for taking a look at this!

In general the implementation looks great! I should have revisited the original issue since working on the object system, because the `Chunk` would now be better off as an implementation of `KotoObject` rather than being a map. No worries if you don't feel like converting it now though, I can always follow up on it later.

Regarding the panic, what you've encountered here is a bug in the runtime that needs to be fixed (_any_ example of the runtime panicking under normal usage should be considered a bug). The problem lies in `KotoVm::run_chunk`: when an error is thrown while running the chunk the execution frame needs to be popped, but it's not which causes any following operations to be working in the wrong frame. I don't have time to push a fix right now, but it's along the lines of adding the following to `run_chunk`:

```
if result.is_err() {
    self.pop_frame(KValue::Null)?;
}
```

File: crates/runtime/src/core_lib/koto.rs [hunk 1]
```diff
@@ -51,5 +55,87 @@ pub fn make_module() -> KMap {
         unexpected => type_error_with_slice("a single argument", unexpected),
     });
 
+    result.add_fn("load", |ctx| match ctx.args() {
+        [KValue::Str(s)] => try_load_koto_script(ctx, s),
+        [KValue::Object(o)] if o.is_a::<File>() => {
+            let mut file = o.cast_mut::<File>().unwrap();
+            let contents = file.read_to_kstring()?;
```
Comment (irh) [inline crates/runtime/src/core_lib/koto.rs:62]: As far as I can tell you don't need a `KString` here, the compiler just wants `&str`, and `KotoFile` returns a `String` from `read_to_string`, so maybe something like a getter for the inner `Ptr<KotoFile>` would do the trick? 
e.g. `let contents = file.inner().read_to_string()?;`?

**Comment by alisomay** [comment]
All sounds good!
I'd gladly convert it to a `KotoObject` and also take a look at the VM bug quickly. 
I'll do the getter for the `File` also.

**Comment by irh** [comment]
Yes this is a bit hairy! One way for you to move forward would be to call `ctx.vm.spawn_shared_vm()` in your function which will give you a usable VM more cheaply than calling `vm.clone()`. 

Otherwise as you've noted, an alternative form of `#[koto_method]` would be needed that clones the instance and args allowing the Vm to passed through as `&mut`. I'm not sure what the ideal design would be there, it might be as simple as just making that the standard behaviour; cloning a `KValue` is intended to be cheap.

However, if this is for the purpose of adding a `.run()` method to compiled chunks, another option would be to not do that, and rely on the standalone `koto.run` function for running chunks, which is maybe nicer anyway (having a single way to do a thing rather than two)?

**Comment by alisomay** [comment]
> However, if this is for the purpose of adding a `.run()` method to compiled chunks, another option would be to not do that, and rely on the standalone `koto.run` function for running chunks, which is maybe nicer anyway (having a single way to do a thing rather than two)?

I think it makes sense to go that way. In terms of errors, `koto.run` would always return a `KValue` either containing the result of the evaluation or the runtime error as a string right? In that case it might be nice to introduce a mechanism to understand if the evaluation errored or not. Maybe we can carry the concept of `chunk.ok` there. 

`koto.run` always return a type named something like `RunResult` or `EvalResult` or a better named one 😅 
So this kind of code can be possible if desired.
```
match koto.load "1 + 1"
  chunk if chunk.ok then 
    match koto.run chunk
      result if result.ok then
         print result
      error then
         throw error
  error then 
    throw error
```
It is a bit nested but if extreme coupling is necessary in terms of errors it could be used. 

> Otherwise as you've noted, an alternative form of #[koto_method] would be needed that clones the instance and args allowing the Vm to passed through as &mut. I'm not sure what the ideal design would be there, it might be as simple as just making that the standard behaviour; cloning a KValue is intended to be cheap.

If that is as cheap as to be negligible I think when necessary it should be trivial to convert the proc macro to generate that version but using only `koto.run` for this specific case sounds cleaner to me. 

`ctx.vm.spawn_shared_vm()` ah cool I didn't know about it. 

For this one, I think I'll go with the `koto.run` idea and see how it feels.

**Comment by alisomay** [comment]
I've made some changes according to our discussion. Let me know how it feels ✨ 

I don't like very much that e.g. `chunk.ok` and `chunk.ok()` are both truthy. 
Because `chunk.ok` would return the method itself and `chunk.ok()` would return a `KValue::Bool`.

Is there a convenient way to still keep a structure like

```rust
#[derive(Clone)]
enum ChunkInner {
    Ok(Ptr<koto_bytecode::Chunk>),
    Err(String),
}

/// The Chunk type used in the koto module
#[derive(Clone, KotoCopy, KotoType)]
pub struct Chunk(ChunkInner);

#[koto_impl(runtime = crate)]
impl Chunk {
    #[koto_method]
    fn ok(&self) -> KValue {
        matches!(self.0, ChunkInner::Ok(_)).into()
    }
}

impl KotoObject for Chunk {
    fn display(&self, ctx: &mut DisplayContext) -> Result<()> {
        match &self.0 {
            ChunkInner::Ok(chunk) => {
                ctx.append(format!("{:?}", chunk));
            }
            ChunkInner::Err(err) => {
                ctx.append(err.to_owned());
            }
        }
        Ok(())
    }
}
```

and mitigate the necessity to call `chunk.ok()` and make `chunk.ok` call it instead?

Or maybe it is confusing.. I think once we decide on a nice API, this is pretty much done 😄

**Comment by irh** [comment]
> I don't like very much that e.g. chunk.ok and chunk.ok() are both truthy.

Yes I agree that's a problem, it would be great if `chunk.ok` was simply a bool, but the `koto_derive` macros currently don't support adding fields to objects, only functions. I'll have a think about how to improve that today.

Another option would be to throw the error from `koto.load`:

```
try
  chunk = koto.load '1 + 1'
catch error
  print error
```

A clue that this might be the way to go is that in the example the error's being thrown anyway =)

`koto.run` could also re-throw runtime errors, which would clear up the meaning of the return type.

**Comment by alisomay** [comment]
Yeah sure! I think this might be a more natural way to use it. I sticked to the initial match usecase. I'll nake the necessary changes today. 👍

About koto.impl, I also think it would be useful to extend the macro to handle fields.

I'd say they're not exposed to koto by default and they get exposed by will by using an attribute macro.

File: crates/derive/src/koto_impl.rs [hunk 2]
```diff
@@ -99,7 +99,7 @@ fn wrap_method(
     let type_name = quote! { #struct_ident::type_static() };
     let fn_name = &sig.ident;
     let fn_ident = quote! {#struct_ident::#fn_name};
-    let wrapper_name = format_ident!("{PREFIX_FUNCTION}{fn_name}");
+    let wrapper_name = format_ident!("{PREFIX_FUNCTION}{struct_ident}_{fn_name}");
```
Comment (alisomay) [inline crates/derive/src/koto_impl.rs:102]: I think this still makes sense to keep uniqueness in struct scope, so we can have identically named methods exposed to koto in the same module but in different structs. 🙏

**Comment by irh** [comment]
>  I haven't fixed the Vm bug yet but I'll get onto it as soon as I can and add it to this branch.

OK I had a moment and pushed a fix.

**Comment by alisomay** [comment]
> @alisomay I've pushed some changes to the branch (I force-pushed after rebasing onto the latest `main` which I shouldn't have done, sorry if that causes you any trouble). I haven't fixed the Vm bug yet but I'll get onto it as soon as I can and add it to this branch.
> 
> While writing docs for the new functions it occurred to me that we could probably get away without the `File` overloads for `load` and `run`. It's simple enough for a script to call `read_to_string` when necessary. What do you think?

Oh no issues at all, I'd just rebase before the merge. 
I think not having a `File` overload is definitely a better idea since a script can basically either be a string or a compiled chunk. `File` seems like an abstraction layer above those. 

I'll remove the overload, update the change log and rebase 👍

**Comment by irh** [comment]
I've merged this now, thanks for taking the time @alisomay!

**Pull request merged**